cpp/src/arrow/dataset/dataset_test.cc

```diff
@@ -58,5 +59,54 @@ TEST_F(TestSimpleDataSource, GetFragments) {
   AssertDataSourceEquals(reader.get(), &source);
 }
 
+TEST(TestProjector, AugmentWithNull) {
+  constexpr int64_t kBatchSize = 1024;
+
+  auto from_schema = schema({field("f64", float64())});
+
+  auto batch = GetRecordBatch(kBatchSize, from_schema);
+
+  auto to_schema = schema({field("i32", int32()), field("f64", float64())});
+
+  RecordBatchProjector projector(default_memory_pool(), to_schema);
+
+  std::shared_ptr<Array> null_i32;
+  ASSERT_OK(MakeArrayOfNull(int32(), batch->num_rows(), &null_i32));
+  auto expected_batch =
+      RecordBatch::Make(to_schema, batch->num_rows(), {null_i32, batch->column(0)});
+
+  std::shared_ptr<RecordBatch> reconciled_batch;
+  ASSERT_OK(projector.Project(*batch, &reconciled_batch));
+
+  AssertBatchesEqual(*expected_batch, *reconciled_batch);
+}
+
+TEST(TestProjector, AugmentWithScalar) {
+  constexpr int64_t kBatchSize = 1024;
+  constexpr int64_t kScalarValue = 3;
+
+  auto from_schema = schema({field("f64", float64())});
+
+  auto batch = GetRecordBatch(kBatchSize, from_schema);
+
+  auto to_schema = schema({field("i32", int32()), field("f64", float64())});
+
+  auto scalar_i32 = std::make_shared<Int32Scalar>(kScalarValue);
+
+  RecordBatchProjector projector(default_memory_pool(), to_schema, {scalar_i32, nullptr});
+
+  std::shared_ptr<Array> array_i32;
+  Int32Builder builder;
+  ASSERT_OK(builder.AppendValues(std::vector<int32_t>(kBatchSize, kScalarValue)));
+  ASSERT_OK(builder.Finish(&array_i32));
+  auto expected_batch =
+      RecordBatch::Make(to_schema, batch->num_rows(), {array_i32, batch->column(0)});
+
+  std::shared_ptr<RecordBatch> reconciled_batch;
+  ASSERT_OK(projector.Project(*batch, &reconciled_batch));
+
+  AssertBatchesEqual(*expected_batch, *reconciled_batch);
+}
+
```

Comment:
I'd add a third test with a lot of re-ordering and subsetting + extra null + extra scalar, e.g:
```
from = {i8, u8, i16, u16, i32, u32}
to = {u32, u16, u8, bool (null), float64(pi)}
```

Reply:
Done